pypesto/objective/priors.py

```diff
@@ -0,0 +1,345 @@
+
+        res = dict()
+
+        for order in sensi_orders:
+            if order == 0:
+                res[FVAL] = self.density_for_full_parameter_vector(x)
+            elif order == 1:
+                res[GRAD] = self.gradient_for_full_parameter_vector(x)
+            elif order == 2:
+                res[HESS] = self.hessian_for_full_parameter_vector(x)
+            else:
+                ValueError(f'Invalid sensi order {order}.')
+
+        return res
+
+    def check_sensi_orders(self,
+                           sensi_orders: Tuple[int, ...],
+                           mode: str) -> bool:
+        for order in sensi_orders:
+            if not (0 <= order <= 2):
+                return False
+        return True
+
+    def check_mode(self, mode) -> bool:
+        if mode == MODE_FUN:
+            return True
+        elif mode == MODE_RES:
+            return False
+        else:
+            ValueError(f'Invalid input: Expected mode {MODE_FUN} or'
+                       f' {MODE_RES}, received {mode} instead.')
+
+    def density_for_full_parameter_vector(self, x):
+
+        density_val = 0
+        for prior in self.prior_list:
+            density_val -= prior['density_fun'](x[prior['index']])
+
+        return density_val
+
+    def gradient_for_full_parameter_vector(self, x):
+
+        grad = np.zeros_like(x)
+
+        for prior in self.prior_list:
+            grad[prior['index']] -= prior['density_dx'](x[prior['index']])
+
+        return grad
+
+    def hessian_for_full_parameter_vector(self, x):
+
+        hessian = np.zeros((len(x), len(x)))
+
+        for prior in self.prior_list:
+            hessian[prior['index'], prior['index']] -= \
+                prior['density_ddx'](x[prior['index']])
+
+        return hessian
+
+    def hessian_vp_for_full_parameter_vector(self, x, p):
+
+        h_dot_p = np.zeros_like(p)
+
+        for prior in self.prior_list:
+            h_dot_p[prior['index']] -= \
+                prior['density_ddx'](x[prior['index']]) * p[prior['index']]
+
+        return h_dot_p
+
+
+def get_parameter_prior_dict(index: int,
+                             prior_type: str,
+                             prior_parameters: list,
+                             parameter_scale: str = 'lin'):
+
+    """
+    Returns the prior dict used to define priors for some default priors.
+
+    index:
+        index of the parameter in x_full
+
+    prior_type: str
+        Prior is defined in LINEAR=untransformed parameter space! prior_type
+        can from {uniform, normal, laplace, logUniform, logNormal, logLaplace}
+
+    prior_parameters:
+        Parameters of the priors. Parameters are defined in linear scale.
+
+    parameter_scale:
+        scale, in which parameter is defined (since a parameter can be
+        log-transformed, while the prior is always defined in the linear space)
+    """
+
+    log_f, d_log_f_dx, dd_log_f_ddx = \
+        _prior_densities(prior_type, prior_parameters)
+
```

Comment:
I think the complexity and maintanance could be substantially improved by defining the prior function once as undefined symbolic function `sympy.Function` and then just using `sympy.diff` and `sympy.lambdify(...,modules=['numpy'])`

Reply:
agreed. that would then need to be cached for efficiency. as now all relevant cases are implemented here and it's only 50loc, this would be fine for me for this PR to get things finished, but potentially for future refactoring.

Reply:
Agreed, fine with me to move on with this PR for now. I don't expect this would need caching  though, lambdify would yield a function that can be evaluated ad hoc. `modules=['numpy']` ensures that this function uses numpy function and we get vectorization etc.

Reply:
yes, I meant to execute the symbolic calculations and lambdification only once at the beginning and not always when calling the function, but that should be fine then.